unix/x0vncserver/XDesktop.cxx

```diff
@@ -383,6 +391,92 @@ KeyCode XDesktop::XkbKeysymToKeycode(Display* dpy, KeySym keysym) {
 }
 #endif
 
+KeyCode XDesktop::addKeysym(Display* dpy, KeySym keysym)
+{
+  unsigned int key;
+  XkbDescPtr xkb;
+  XkbMapChangesRec changes;
+  KeySym *syms;
+  xkb = XkbGetMap(dpy, XkbAllComponentsMask, XkbUseCoreKbd);
+
+  if (!xkb)
+    return false;
+
+  for (key = xkb->max_key_code; key >= xkb->min_key_code; key--) {
+    if (XkbKeyNumGroups(xkb, key) == 0)
+      break;
+  }
+
+  if (key < xkb->min_key_code)
+    return false;
+
+  memset(&changes, 0, sizeof(changes));
+
+  syms = XkbKeySymsPtr(xkb, key);
+  syms[0] = keysym;
```

Comment:
You sure this order is safe? I'm thinking that there might not be any space allocated before `XkbChangeTypesOfKey()` is called.